src/datafusion/file_opener.rs

```diff
@@ -0,0 +1,105 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow_schema::ArrowError;
+use datafusion::{datasource::physical_plan::FileOpener, error::DataFusionError};
+use futures::{StreamExt, TryStreamExt};
+
+use crate::async_reader::{ZarrPath, ZarrRecordBatchStreamBuilder};
+
+use super::config::ZarrConfig;
+
+pub struct ZarrFileOpener {
+    config: ZarrConfig,
+}
+
+impl ZarrFileOpener {
+    pub fn new(config: ZarrConfig) -> Self {
+        Self { config }
+    }
+}
+
+impl FileOpener for ZarrFileOpener {
+    fn open(
+        &self,
+        file_meta: datafusion::datasource::physical_plan::FileMeta,
+    ) -> datafusion::error::Result<datafusion::datasource::physical_plan::FileOpenFuture> {
+        let config = self.config.clone();
+
+        Ok(Box::pin(async move {
+            let zarr_path = ZarrPath::new(config.object_store, file_meta.object_meta.location);
+
+            let rng = file_meta.range.map(|r| (r.start as usize, r.end as usize));
+            let batch_reader = ZarrRecordBatchStreamBuilder::new(zarr_path)
+                .build_partial_reader(rng)
+                .await
+                .map_err(|_| {
+                    DataFusionError::Execution("Error creating zarr reader".to_string())
+                })?;
+
+            let stream = batch_reader
+                .map_err(|_| ArrowError::ComputeError("Error reading zarr".to_string()));
+
+            Ok(stream.boxed())
+        }))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use datafusion::datasource::physical_plan::FileMeta;
+    use object_store::{local::LocalFileSystem, path::Path, ObjectMeta};
+
+    use crate::tests::get_test_v2_data_path;
+
+    use super::*;
+
+    #[tokio::test]
+    async fn test_open() {
+        let local_fs = LocalFileSystem::new();
+
+        let test_data = get_test_v2_data_path("lat_lon_example.zarr".to_string());
+
+        let config = ZarrConfig {
+            object_store: Arc::new(local_fs),
+        };
+
+        let opener = ZarrFileOpener::new(config);
+
+        let file_meta = FileMeta {
```

Comment:
I'm a little worried about how `lat_lon_example.zarr` actually being a directory will cause challenges downstream. E.g. `object_store.head(&path)` will throw an error, which is why this is manually constructed.

Reply:
Oh you mean that the datafusion concept of a file meta isn't well adapted for a directory? Does the Opener struct have to work with a file meta (the one defined in datafusion)? I guess we should keep this as close as possible to other datafusion file opener implementations, but then, the fact that Zarr doesn't exactly expose a single file that can be atomically consumed (e.g. different variables are stored across different files) means it's pretty different from a parquet or csv... In any case, sounds like a slightly more manual way of constructing the file meta is a fine work around for now?

Reply:
Yeah, it doesn't really have a concept of treating a directory as having different parts for a datasets. It does work well with lists of files (like a bunch of parquet files), but it mostly lists them, passes them to the `FileOpener::open` method, and coalesces them. So e.g., not sure how this would work: https://docs.rs/datafusion/latest/datafusion/datasource/listing/struct.ListingTable.html

I think things are fine for now (constructing manually), but we'll see which issues we hit and if we can work around them hear, or maybe opportunities to improve DataFusion upstream.

Reply:
Right right, and I read the code too fast, I see now that you're implementing a trait from `datafusion`, `FileOpener`, which exposes a method that takes in a file meta, so of course "the Opener struct has to work with the file meta", the trait requires it. Okay let's just go ahead with this for now.